Update quickstart tutorials and acknowledgements (#307)
learning-at-home/hivemind: docs/user/moe.md

# Mixture-of-Experts

> This tutorial covers the basics of Decentralized Mixture of Experts (DMoE).

Reviewer: ```suggestion
This tutorial covers the basics of Decentralized Mixture-of-Experts (DMoE).
```

From the infrastructure standpoint, DMoE consists of two parts: experts hosted on peer devices, and a gating/routing function that assigns input to one of these experts.

## Host experts with a server

`hivemind.moe.Server` hosts one or several experts (PyTorch modules) for remote access. These experts are responsible for
most of the model parameters and computation. The server can be started using either Python or
[a shell script](https://github.com/learning-at-home/hivemind/blob/master/hivemind/hivemind_cli/run_server.py). We'll use the shell
for now. To host a server with default experts, run this in your shell:

```sh
hivemind-server --expert_cls ffn --hidden_dim 512 --num_experts 5 --expert_pattern "expert.[0:5]" \
                --listen_on 0.0.0.0:1337 --dht_port 1338
# note: if you omit listen_on and/or dht_port, they will be chosen automatically and printed to stdout.
```

<details style="margin-top:-24px; margin-bottom: 16px;">
  <summary><i>Console outputs</i></summary>

```sh

> [2021/07/15 18:52:01.424][INFO][moe.server.create:156] Running DHT node on ['/ip4/127.0.0.1/tcp/42513/p2p/QmacLgRkAHSqdWYdQ8TePioMxQCNV2JeD3AUDmbVd69gNL'], initial peers = []

Reviewer: The port differs from 1338, are you sure this is the correct output?

Author: fixd

[2021/07/15 18:52:02.484][INFO][moe.server.task_pool.run:145] expert.0_backward starting, pid=14051
[2021/07/15 18:52:02.501][INFO][moe.server.task_pool.run:145] expert.3_forward starting, pid=14057
[2021/07/15 18:52:02.508][INFO][moe.server.task_pool.run:145] expert.3_backward starting, pid=14058
[2021/07/15 18:52:02.508][INFO][moe.server.task_pool.run:145] expert.2_forward starting, pid=14060
[2021/07/15 18:52:02.521][INFO][moe.server.task_pool.run:145] expert.2_backward starting, pid=14070
[2021/07/15 18:52:02.521][INFO][moe.server.task_pool.run:145] expert.1_forward starting, pid=14075
[2021/07/15 18:52:02.532][INFO][moe.server.task_pool.run:145] expert.1_backward starting, pid=14081
[2021/07/15 18:52:02.532][INFO][moe.server.runtime.run:80] Started
```

</details>

This server serves 5 feedforward experts with ReLU and LayerNorm
(see
architecture [here](https://github.com/learning-at-home/hivemind/blob/master/hivemind/server/layers/__init__.py#L7-L21))
. In order to connect to this server, you should copy its address from console outputs:
```shell
[...][INFO][moe.server.create:156] Running DHT node on ['ADDRESS_WILL_BE_PRINTED_HERE']
```

> You can create additional servers in the same decentralized network using `--initial_peers` argument:

Reviewer: ```suggestion
You can create additional servers in the same decentralized network using the `--initial_peers` argument:
```

[2021/07/15 18:53:44.058][INFO][moe.server.task_pool.run:145] expert.106_backward starting, pid=14919
[2021/07/15 18:53:44.077][INFO][moe.server.task_pool.run:145] expert.31_forward starting, pid=14923
[2021/07/15 18:53:44.077][INFO][moe.server.task_pool.run:145] expert.31_backward starting, pid=14925
[2021/07/15 18:53:44.095][INFO][moe.server.task_pool.run:145] expert.95_forward starting, pid=14932
[2021/07/15 18:53:44.106][INFO][moe.server.task_pool.run:145] expert.95_backward starting, pid=14935
[2021/07/15 18:53:44.118][INFO][moe.server.task_pool.run:145] expert.167_forward starting, pid=14943
[2021/07/15 18:53:44.119][INFO][moe.server.task_pool.run:145] expert.167_backward starting, pid=14944
[2021/07/15 18:53:44.123][INFO][moe.server.runtime.run:80] Started
```

</details>

By default, the server will only accept connections from your local machine. To access it globally, you should replace
`127.0.0.1` part from initial peers with server's IP address. Hivemind supports both ipv4 and ipv6 protocols and uses the same notation
as [libp2p](https://docs.libp2p.io/concepts/addressing/). You can find more details on multiaddresses in the 

> [DHT tutorial](https://learning-at-home.readthedocs.io/en/latest/user/dht.html)ю 

Reviewer: ```suggestion
[DHT tutorial](https://learning-at-home.readthedocs.io/en/latest/user/dht.html).
```

## Train the experts

Now let's put these experts to work. Create a python console (or a jupyter) and run:

```python
import torch
import hivemind

dht = hivemind.DHT(
    initial_peers=["/ip4/127.0.0.1/tcp/TODO/COPYFULL_ADDRESS/FROM_ONE_OF_THE_SERVERS"],
    client_mode=True, start=True)

> # note: listen=False means that your peer will operate in "client only" mode: 

Reviewer: ```suggestion
# note: client_mode=True means that your peer will operate in a "client-only" mode: 
```

# this means that it can request other peers, but will not accept requests in return 

expert1, expert4 = hivemind.moe.get_experts(dht, ["expert.1", "expert.4"])
assert expert1 is not None and expert4 is not None, "experts not found. Please double-check initial peers"
```

> The experts (e.g. `expert1`) can be used as a pytorch module with autograd support:

Reviewer: ```suggestion
Each expert (e.g. `expert1`) can be used as a pytorch module with autograd support:
```

```python
dummy = torch.randn(3, 512)
out = expert1(dummy)  # forward pass
out.sum().backward()  # backward pass
```

> When called, expert1 will submit a request to the corresponding server (which you created above) and return the output

Reviewer: ```suggestion
When called, `expert1` will submit a request to the corresponding server (which you created above) and return the output
```
